Process block headers

samay-kothari opened the pull request:
Using the functions introduced in #22 to add the functionality where we would be processing headers just after they are downloaded.

samay-kothari commented:
@kcalvinalvin can you rerun the test, I don't know why the check is failed here, it passed on my forked repository.
https://github.com/samay-kothari/utreexod/actions/runs/2798870092

kcalvinalvin commented:
> @kcalvinalvin can you rerun the test, I don't know why the check is failed here, it passed on my forked repository.
> https://github.com/samay-kothari/utreexod/actions/runs/2798870092

Ok. For future reference, you can force github to re-run actions by changing the commit hash with `git commit --amend` and force pushing.

kcalvinalvin commented on blockchain/blockindex.go:
```diff
@@ -269,7 +269,8 @@ func newBlockIndex(db database.DB, chainParams *chaincfg.Params) *blockIndex {
 func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
 	bi.RLock()
-	_, hasBlock := bi.index[*hash]
+	node := bi.LookupNode(hash)
```
`bi.LookupNode()` also tries to acquire the blockindex read lock. It'll work because it's a read lock but it ideally should be avoided. I think it's better to use `bi.index[*hash]`.

kcalvinalvin commented on blockchain/blockindex.go:
```diff
@@ -269,7 +269,8 @@ func newBlockIndex(db database.DB, chainParams *chaincfg.Params) *blockIndex {
 	bi.RLock()
-	_, hasBlock := bi.index[*hash]
+	node := bi.LookupNode(hash)
+	hasBlock := node != nil && node.status.HaveData()
```
This line changes the behavior of the function `HaveBlock()` the documentation states that it returns whether a block exists in the `blockindex`. This line also asserts that the blockdata is also downloaded.

HaveBlock is only called in `blockchain.blockExists` in this repo anyways so why not edit that function instead? `HaveBlock` is exported so best not to change its behavior unless there's no alternative for backwards compatibility (for potential people using the library).

samay-kothari replied:
Yeah this makes sense
I will make the required changes to blockExists and test the functionality

kcalvinalvin commented on blockchain/process.go:
```diff
@@ -202,7 +202,7 @@ func (b *BlockChain) ProcessBlock(block *btcutil.Block, flags BehaviorFlags) (bo
 
 	// The block must not already exist in the main chain or side chains.
-	exists, err := b.blockExists(blockHash)
+	exists, err := b.HaveBlock(blockHash)
```
Note the comment on `HaveBlock`. We can change the behavior of `blockExists` instead of changing the behavior of `HaveBlock`

samay-kothari replied:
I have changed the behaviour of blockExists and it checks and returns, whether there is block data or not

samay-kothari commented on blockchain/process.go:
```diff
@@ -42,7 +42,8 @@ const (
+		return true && node.status.HaveData(), nil
 	}
 
 	// Check in the database.
```
@kcalvinalvin in the blockExists function, do I need to check whether block data exists for the block stored in database or not. Can a node be stored in the database, just with its header information?

kcalvinalvin commented on netsync/manager.go:
```diff
@@ -1008,7 +1009,17 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
+			return
+		}
 		// Verify the header at the next checkpoint height matches.
 		if node.height == sm.nextCheckpoint.Height {
```
We wouldn't need this check here because it's redundant with the checkpoint check in `b.checkHeaderContext` that's called in `maybeAcceptBlockHeader`, which is called from `ProcessBlockHeader`.

https://github.com/utreexo/utreexod/blob/53dc2e1096418f7286cb29e7d0b679c3129a524a/blockchain/validate.go#L676-L697

kcalvinalvin commented on netsync/manager.go:
```diff
@@ -1008,7 +1009,17 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
-
+		// Processing the block headers that are downloaded and if it passes
+		// all the checks, creating a block node that only contains header.
+		err := chain.ProcessBlockHeader(blockHeader)
```
If we call this, then we'd also need to get rid of the `checkBlockHeaderContext` call in `checkBlockContext`, which is called from `maybeAcceptBlock` since it's just duplicate check of the same exact thing.

https://github.com/utreexo/utreexod/blob/53dc2e1096418f7286cb29e7d0b679c3129a524a/blockchain/validate.go#L726-L732

samay-kothari commented:
Hi calvin, I was wondering that the usage of blockExists here https://github.com/utreexo/utreexod/blob/53dc2e1096418f7286cb29e7d0b679c3129a524a/blockchain/process.go#L266-L271, where we are checking if the block is orphan or not.

Ideally we should not add the condition of presence of block data along with presence of node while checking if the parent exists or not because the case might be in future, that the parent's header is verified, but the block is not downloaded yet.

So I think rather than changing the blockExists function, I can add a check for block data wherever it is  required.

samay-kothari commented:
Hi calvin, as discussed in the call, i added the flags that prevents the header validation to happen if already the header is verified. I also removed the changes made to the existing `blockExists` function, and added the new function that checks if the block is present in data, so existing functioning is not getting affected.

samay-kothari commented on blockchain/process.go:
```diff
@@ -200,13 +200,20 @@ func (b *BlockChain) ProcessBlock(block *btcutil.Block, flags BehaviorFlags) (bo
 	blockHash := block.Hash()
 	log.Tracef("Processing block %v", blockHash)
-
 	// The block must not already exist in the main chain or side chains.
```
Yeah that makes, does `Looking up for node and checking if it has block data or not` works?

samay-kothari commented on blockchain/process.go:
```diff
@@ -200,13 +200,20 @@ func (b *BlockChain) ProcessBlock(block *btcutil.Block, flags BehaviorFlags) (bo
 		return false, false, err
 	}
-	if exists {
+	// Looking up for node and checking if it have block data or not
```
I will change `have` with `has`. :)

samay-kothari commented on blockchain/accept.go:
```diff
@@ -62,23 +62,29 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
-	err = b.index.flushToDB()
-	if err != nil {
-		return false, err
+	blockHash := blockHeader.BlockHash()
```
Okay will make these changes

samay-kothari commented on blockchain/process.go:
```diff
@@ -33,6 +33,11 @@ const (
+	// BFHeaderValidated sets the third bit of the flag and indicates that
+	// the header has already been validated and the header validation can be
+	// skipped when block validation is being performed.
+	BFHeaderValidated BehaviorFlags = 1 << 2
```
Okay I will do that

samay-kothari commented on blockchain/validate.go:
```diff
@@ -467,9 +467,15 @@ func checkBlockHeaderSanity(header *wire.BlockHeader, powLimit *big.Int, timeSou
+	// If the third bit of the flag is set, then it means that the header
+	// is already been validated, so we don't need to revalidate it. So we
+	// check block header sanity, only when the second bit is not set to 1
+	if (flags >> 2 & 1) != 1 {
```
Yeah this is more readable and understandable

kcalvinalvin reviewed (changes requested):
Could you also wrap the commit messages at 70?

Ex:

```
When the block data is being validated for an already validated header, then we modify the behavioural flag, so that we don't repeat the header validation steps again
```

should be 

```
When the block data is being validated for an already validated
header, then we modify the behavioural flag, so that we don't repeat
the header validation steps again
```

Also am seeing some grammatical errors here and there and it'd be better to have them grammatically correct.

samay-kothari commented on blockchain/chain.go:
```diff
@@ -228,6 +228,39 @@ func (b *BlockChain) HaveBlock(hash *chainhash.Hash) (bool, error) {
+	if exists || b.IsKnownOrphan(hash) {
+		// Retrieving node from the blockIndex so that we can check the node
+		// status.
+		node := b.IndexLookupNode(hash)
```
I am checking again for the node in `IndexLookupNode` because of the case where the block is stored in the database and is not present in the Index.
In that case the returned node would be nil right?

samay-kothari replied:
I got your point, that checking for block data for an orphan block has no point, since we can't have an orphan header.
So I can just return true if the block is an orphan.

samay-kothari commented on netsync/manager.go:
```diff
@@ -1008,7 +1008,17 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
+				"disconnecting", blockHeader.BlockHash(), peer, err)
+			peer.Disconnect()
+			return
+		}
```
On line 1023 the check for `node.height == sm.nextCheckpoint.Height`, is to check that if it is a checkpoint then break the loop over headers, which I think is necessary.

I think the check `node.hash.IsEqual(sm.nextCheckpoint.Hash)` at line 1024 is redundant and I can remove the if-else and output and process the if condition only.

kcalvinalvin commented on blockchain/fullblocktests/generate.go:
```diff
@@ -2317,11 +2342,79 @@ func GenerateHeaders() (generator *testGenerator, tests [][]TestInstance) {
-	g.nextBlock("b4", outs[1])
+	g.nextBlock("b4", outs[2])
 	accepted()
+	// Adding a block with valid header and valid spend, but invalid parent
```
The comment is wrong here. In headers-first you don't accept a header with an invalid parent. The tip is never reset to `"b0"` and thus it has a valid parent of `"b4"`.

In general, checkpointed headers are assumed to be correct as it's hardcoded into the binary and it's actually a consensus rule (whether or not this is good is besides the point). Once you accept headers, the block **must** match with the header. If it doesn't I believe the node just hangs.

So we can have tests for rejecting headers but once we accept the headers, it's assumed as correct so the blockdata corresponding to the headers should never be rejected.
